### PR TITLE
feat: use ws subscriptions (`eth_subscribe`) for `watchBlocks`, `watchBlockNumber` & `watchPendingTransactions`

### PR DESCRIPTION
**Pull request by jxom**
Added WebSocket `eth_subscribe` support `watchBlocks`, `watchBlockNumber`, and `watchPendingTransactions`.

### CODE REVIEW

File: src/actions/public/watchBlockNumber.test.ts [hunk 1]
```diff
@@ -5,283 +5,427 @@ import { watchBlockNumber } from './watchBlockNumber'
 import { publicClient, testClient } from '../../_test'
 import { wait } from '../../utils/wait'
 import { localhost } from '../../chains'
-import { createPublicClient, http } from '../../clients'
+import { createPublicClient, http, PublicClient } from '../../clients'
 import { mine } from '../test/mine'
 import * as getBlockNumber from './getBlockNumber'
 import { setIntervalMining } from '../test'
+import { webSocketClient } from '../../_test/utils'
```
Comment (tmm) [inline src/actions/public/watchBlockNumber.test.ts:12]: Should we export `webSocketClient` from `src/_test`?

File: site/docs/actions/public/watchBlockNumber.md [hunk 2]
```diff
@@ -95,6 +95,32 @@ const unwatch = publicClient.watchBlockNumber(
 )
 ```
 
+### poll (optional)
+
+- **Type:** `boolean`
+- **Default:** `false` for WebSocket Clients, `true` for non-WebSocket Clients
```
Comment (tmm) [inline site/docs/actions/public/watchBlockNumber.md:101]:
```suggestion
- **Default:** `false` for WebSocket Transports, `true` for non-WebSocket Transports
```

File: site/docs/actions/public/watchPendingTransactions.md [hunk 3]
```diff
@@ -116,5 +142,7 @@ const unwatch = publicClient..watchPendingTransactions(
 
 ## JSON-RPC Methods
 
-- Calls [`eth_newPendingTransactionFilter`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_newpendingtransactionfilter) to initialize the filter.
-- Calls [`eth_getFilterChanges`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getFilterChanges) on a polling interval.
\ No newline at end of file
+- When `poll: true`
+  - Calls [`eth_newPendingTransactionFilter`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_newpendingtransactionfilter) to initialize the filter.
+  - Calls [`eth_getFilterChanges`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getFilterChanges) on a polling interval.
+- When `poll: false` & WebSocket Transport, uses a WebSocket subscription via [`eth_subscribe`](https://docs.alchemy.com/reference/eth-subscribe-polygon) and the `"newPendingTransactions"` event.
```
Comment (tmm) [inline site/docs/actions/public/watchPendingTransactions.md:148]: Great comment 🫡

File: src/actions/public/waitForTransactionReceipt.ts [hunk 4]
```diff
@@ -77,6 +77,7 @@ export async function waitForTransactionReceipt<
         const unwatch = watchBlockNumber(client, {
           emitMissed: true,
           emitOnBegin: true,
+          poll: true,
```
Comment (tmm) [inline src/actions/public/waitForTransactionReceipt.ts:80]: Should `waitForTransactionReceipt` have an option to "subscribe"?

File: src/actions/public/watchBlockNumber.test.ts
Comment (tmm) [inline src/actions/public/watchBlockNumber.test.ts:1]: `poll: false` test for webSocketClient could be nice at some point

File: src/actions/public/watchBlockNumber.ts [hunk 5]
```diff
@@ -11,77 +12,129 @@ export type OnBlockNumberFn = (
...
-}
+} & (GetTransportConfig<TTransport>['type'] extends 'webSocket'
+  ?
+      | {
+          emitMissed?: never
+          emitOnBegin?: never
+          /** Whether or not the WebSocket Transport should poll the JSON-RPC, rather than using `eth_subscribe`. */
+          poll?: false
+          pollingInterval?: never
+        }
```
Comment (tmm) [inline src/actions/public/watchBlockNumber.ts:39]: Could break out this into a utility type at some point.

File: src/actions/public/watchBlockNumber.ts [hunk 6]
```diff
@@ -11,77 +12,129 @@ export type OnBlockNumberFn = (
...
+      | (PollOptions & { poll: true })
+  : PollOptions & { poll?: true })
```
Comment (tmm) [inline src/actions/public/watchBlockNumber.ts:41]: Should we remove `poll` as an option in this case?

```suggestion
  : PollOptions & unknown)
```

Or setting it to `undefined`

```suggestion
  : PollOptions & { poll?: undefined })
```

File: src/actions/public/watchBlocks.ts [hunk 7]
```diff
@@ -1,52 +1,52 @@
...
-export type OnBlock<
-  TChain extends Chain | undefined = Chain | undefined,
-  TIncludeTransactions = false,
-> = (
-  block: OnBlockParameter<TChain, TIncludeTransactions>,
-  prevBlock: OnBlockParameter<TChain, TIncludeTransactions> | undefined,
+export type OnBlockParameter<TChain extends Chain | undefined = Chain> =
+  GetBlockReturnType<TChain>
+
+export type OnBlock<TChain extends Chain | undefined = Chain,> = (
```
Comment (tmm) [inline src/actions/public/watchBlocks.ts:11]:
```suggestion
export type OnBlock<TChain extends Chain | undefined = Chain> = (
```

File: src/actions/public/watchBlocks.ts [hunk 8]
```diff
@@ -55,77 +55,111 @@ export type WatchBlocksReturnType = () => void
  */
 export function watchBlocks<
   TChain extends Chain | undefined,
-  TWatchBlocksParameters extends WatchBlocksParameters<TChain>,
+  TTransport extends Transport,
```
Comment (tmm) [inline src/actions/public/watchBlocks.ts:58]: Let's keep `TTransport` in the first generic slot for actions.

**Review by tmm — APPROVED** [review]
few things to consider

File: src/actions/public/waitForTransactionReceipt.ts (hunk 4 above)
Comment (jxom) [reply src/actions/public/waitForTransactionReceipt.ts:80]: WS subscriptions are a bit unreliable here unfortunately. When I tested using Anvil, some block numbers were missed.

File: src/actions/public/watchBlockNumber.test.ts
Comment (jxom) [reply src/actions/public/watchBlockNumber.test.ts:1]: We already do when we run the tests in WebSocket mode :D

Comment (tmm) [reply src/actions/public/watchBlockNumber.test.ts:1]: lol obv

**Pull request merged**